### ISSUE TITLE
Delay task enqueue

### ISSUE THREAD

**Issue by taylor1791** [Nukesor/pueue#72]
I often need to schedule certain tasks. e.g. `scp` this file tomorrow. Having this feature built in would be really nice.

```
pueue add -- scp file user@host:/srv/www/releases/
pueue stash <task-id>
pueue enqueue --delay-until '2020-04-01T00:00:00' <task-id>
```

If combined with the tags feature mentioned in !44, one could have a finer control over when a task was actually completed. Imagine creating a tag for short lived, high urgency tasks. 

I have been working around it currently, by using `sleep 3600 && pueue add ...` in a tmux session, but having 7 of those (one for every day of the week is really annoying). I have tried using `at`, but I can't do that because it requires disabling SIP on MacOS these days. Ug.

By adding this feature it might also require being able to add a stashed task. I could imagine this being `pueue add --stash -- ...`. This would avoid executing it immediately if the queue happened to be empty.

To create a really nice user experience, an option like `--delay-until now + 2 days` or `--delay-until tomorrow` would be really nice. This syntax was taken from the `at` command. That being said, I would be more than happy with any format.

I looked into the source and I am willing and feel pretty confident I could make the contribution after agreeing on a design.

**Comment by Nukesor** [comment]
I really like the idea. It should have a quite low impact on the code base as well.

Until the feature is added, you can simply do `pueue add "sleep 3600 && your_command"`. This should remove the necessity for tmux.

1. A `--stash` flag would even be convenient without the delay feature.
2. Adding `--delay-until` it to `add` and `enqueue` sounds good :). I agree that having tags would be super convenient to ensure that the task will be started at this point of time. Alternatively one could also add a `--start-at` flag, which forces the start of the task at a specific point of time. Although, this might be something for another PR.
2.1. However, I don't know yet how this should be shown in the UI. Another column which is only displayed, if there is at leas one delayed task?  
3. For parsing the input, [humantime](https://github.com/tailhook/humantime) looks like a good solution. I really don't want to add the overhead of maintaining a custom written parser to Pueue.

Feel free to start hacking :)! I'll try to answer all questions you might have

**Comment by taylor1791** [comment]
Here is the plan, broken into decisions. Let me know what you think. My potential concerns are listen at the end.

1. **`pueue add`**. Add `--delay <str>` to `add`. It will require the `--stash` option.
1. **`pueue enqueue`**. Add `--delay <str>` argument.
1. **`--delay` parsing**. Use the first `Ok` Result from `humantime::parse_rfc3339_weak`, `humantime::parse_duration`, and `std::str::FromStr::from_str()`. Durations will be treated relative to relative to `Local::now()`. `from_str` will assume a duration in seconds
1. **`pueue status`** Display the column "Enqueue In" to the left of "status" if relevant exists. Each cell will be blank or have a duration formatted with `humantime::format_rfc3339`. 
1. **Task Update** Replace `Stashed` in the `TaskStatus` enum to `Stashed { enqueue_at: DateTime<Local> }`. There are implication to `PartialEq` here. Replace the uses of `PartialEq` with function calls.
1. **Task Handler** Define `self.processed_stashed` which will filter the tasks down to stashed tasks with an `enqueue_at` and change the status to queued if the time is past.

I am a little nervous about the last step since it is the first thing to do O(tasks) work in the main loop. With sufficiently many tasks, this would cause problems. However, the client doesn't work well with that many tasks either, so I am suspecting you are okay with that choice.

**Comment by Nukesor** [comment]
1. If `--delay` is given, `--stash` should be implied. No need to make the UI more complicated than it has to be.
2. Sounds good. I'm not sure yet, if the parameter should be `delay` as well, since it's kind of another operation (delaying a start vs. scheduling from a stashed state). On the other hand, having a consistent syntax is nice as well.... Let's stick with `--delay` for now on both commands.
3. Sounds good as well. I'm curious how well this will work in the end. I guess we'll need to add extensive docs to describe which syntax users can use for this option.
4. Yep. It should probably be `Enqueue at`, though.
5. I would love to keep the TaskStatus Enum simple without any additional logic. What about adding `enqueue_at: Option<DateTime<Local>>` to `Task`? This will also prevent any problems with `PartialEq`.
6. Yet another naming thing :laughing: . How about `self.check_stashed`? I might be wrong. In my understanding `processing` takes something and transforms it into another thing. E.g. a task into a finished task, which are two quite distinct things on a logical level. Checking if something is ready and changing a flag doesn't feel significant enough for a `processing`. Not sure If this makes any sense, I'm not a native speaker after all :sweat_smile:. 
Please correct me, if I'm wrong :)

**Comment by taylor1791** [comment]
I have to admit I made a mistake in my description of 4. I meant to say `humantime::format_duration` with the column titled `Enqueue in`. I picked duration over a timestamp because of its immediacy by avoiding mental calculations. In addition, the column is usually smaller than a 21 character timestamp and can be made even smaller by showing the most 2 significant units e.g. `3d 1h` instead of `3d 1h 4m 1s 43ms`. While making this decision I referenced [UX Movement](https://uxmovement.com/content/absolute-vs-relative-timestamps-when-to-use-which/).

Despite my obvious preference to duration, I'll implement whichever one you decide.

**Comment by taylor1791** [comment]
I began to work on parsing `--delay`. `humantime` is pretty restrictive with their implementation of `parse_rfc3339_weak`. Here are some of my complaints:

 1) Uses std::time::Duration instead of chrono::Duration
 1) Requires leading zeros
 1) Requires the timestamp
 1) Requires seconds on the timestamp

After looking around I discovered `chrono_english`. It solves all my complaints and accepts all of the formats bellow and more for "free." It is even less code! Do you mind if we use `chrono_english` instead?

```
2020-04-01
1 April 2020
April 1, 2020
04/01/20
April 1
next April 1
next friday
2 days 1min
18:03
6.03pm
next friday 8pm
```

**Comment by Nukesor** [comment]
`chrono_english` looks surprisingly good. I'm surprised it didn't get more attention yet.
Feel free to use it :)

To showing absolute vs. relative times:
IMO, when scheduling a task for later execution, one usually has a certain point in time in mind.
something like: I want this to run in two days at 1pm.

I imagine that showing the relative time will result in a higher cognitive overhead in most of the times. Usually one's interested in **when** the command will be scheduled, rather than how much time will pass until it'll be scheduled.

If the width of the column is too long, one could simply split the date and time with a newline.
Horizontal scrolling and place isn't a problem for most people. E.g.
```
2020-03-02
13:05:12
```

**Comment by taylor1791** [comment]
Just an update.

I have the first pass implementation complete. In addition to my manual testing, I am going to run it locally for a few days before I submit the MR.

**Comment by Nukesor** [comment]
Feel free to already create the PR :)
This will give me some time to review your code. Maybe I can already find some issues or suggest improvements!

**Comment by taylor1791** [comment]
This has been implemented #75. A new version with these features will be published in the near future.
